File: src/hw/vga.c
/* vga.c -- simple VGA library.

   This is free and unencumbered software released into the public domain.
   For more information, please refer to <http://unlicense.org>. */

#include "defines.h"

#ifdef __WATCOMC__
#pragma aux default "$hw$vga$*"
#endif

#include <stdint.h>
#include "pascal.h"
#include "cc/i86.h"
#include "cc/conio.h"
#include "cc/stdlib.h"
#include "cc/string.h"
#include "debug.h"
#include "hw/vbios.h"
#include "hw/vga.h"

/*
 * "vid_port" - port address for 6845 video controller chip (0x3B4 or 0x3D4).
 */
void vga_wait_sync (uint16_t vid_port, bool disable_irqs)
{
    uint16_t io;
    io = vid_port + 6;  /* port 0x3da: read status register */
    /* bit 0 - if set, display is in vertical or horizontal retrace */
    while (inp (io) & 1);
    if (disable_irqs) _disable ();
    while (!(inp (io) & 1));
}

/*
 * "vid_port" - port address for 6845 video controller chip (0x3B4 or 0x3D4).
 */
void vga_wait_vsync (uint16_t vid_port, bool disable_irqs)
{
    uint16_t io;
    io = vid_port + 6;  /* port 0x3da: read status register */
    /* bit 3 - if set, display is in vertical retrace */
    while (!(inp (io) & 8));
    if (disable_irqs) _disable ();
    while (inp (io) & 8);
}

/*
 * "vid_port" - port address for 6845 video controller chip (0x3B4 or 0x3D4).
 */
void vga_set_text_cursor_position (uint16_t vid_port, uint16_t offset)
{
    /* port 0x3d4: CRT controller register select (write) */
    /* port 0x3d5: CRT controller register read/write */
    outp (vid_port + 0, 0x0e);  /* 0x0e - cursor position (MSB) */
    outp (vid_port + 1, offset >> 8);
    outp (vid_port + 0, 0x0f);  /* 0x0f - cursor position (LSB) */
    outp (vid_port + 1, offset & 0xff);
}

void vga_clear_page_320x200x8 (void *buf, char color)
{
    memset (buf, color, 320*200);
}

/*** Initialization ***/

void __near vga_init(void)
{
}

void __near vga_done(void)
{
}

DEFINE_REGISTRATION(vga, vga_init, vga_done)

File: src/hw/vga.pas
(* vga.pas -- declarations for vga.c.

   This file is for linking compiled object files with Pascal linker.
   It will be deleted in future when we rewrite the project in C.

   This is free and unencumbered software released into the public domain.
   For more information, please refer to <http://unlicense.org>. *)

unit vga;

interface

(*$I defines.pas*)

procedure vga_wait_sync;
procedure vga_wait_vsync;
procedure vga_set_text_cursor_position;
procedure vga_clear_page_320x200x8;
procedure vga_line;
procedure vga_bar;

procedure register_vga;

implementation

uses
    pascal,
    i86,
    stdlib,
    string_,
    conio,
    debug;

(*$l vga.obj*)

procedure vga_wait_sync; external;
procedure vga_wait_vsync; external;
procedure vga_set_text_cursor_position; external;
procedure vga_clear_page_320x200x8; external;

(*$L vga\bar.obj*)
procedure vga_bar; external;

(*$L vga\line.obj*)
procedure vga_line; external;

procedure register_vga; external;

end.
